internals/overlord/servstate/manager_test.go

```diff
@@ -902,91 +690,74 @@ PEBBLE_ENV_TEST_PARENT=from-parent
 }
 
 func (s *S) TestActionRestart(c *C) {
-	s.setupDefaultServiceManager(c)
+	s.newServiceManager(c)
+	s.planAddLayer(c, testPlanLayer)
+
 	// Add custom backoff delay so it auto-restarts quickly.
-	layer := parseLayer(c, 0, "layer", `
+	s.planAddLayer(c, `
 services:
     test2:
         override: merge
-        command: /bin/sh -c "echo test2; exec sleep 10"
-        backoff-delay: 50ms
-        backoff-limit: 150ms
+        command: /bin/sh -c "echo test2; {{.NotifyDoneCheck}}; sleep 10"
+        backoff-delay: 1ms
+        backoff-limit: 500ms
+        backoff-factor: 10
 `)
-	err := s.manager.AppendLayer(layer)
-	c.Assert(err, IsNil)
+	s.planChanged(c)
 
 	// Start service and wait till it starts up the first time.
 	chg := s.startServices(c, []string{"test2"}, 1)
-	s.waitUntilService(c, "test2", func(svc *servstate.ServiceInfo) bool {
-		return svc.Current == servstate.StatusActive
-	})
+	s.waitForDoneCheck(c, "test2")
 	c.Assert(s.manager.BackoffNum("test2"), Equals, 0)
 	s.st.Lock()
 	c.Check(chg.Status(), Equals, state.DoneStatus)
 	s.st.Unlock()
-	time.Sleep(10 * time.Millisecond) // ensure it has enough time to write to the log
 	c.Check(s.readAndClearLogBuffer(), Matches, `2.* \[test2\] test2\n`)
 
 	// Send signal to process to terminate it early.
-	err = s.manager.SendSignal([]string{"test2"}, "SIGTERM")
+	err := s.manager.SendSignal([]string{"test2"}, "SIGTERM")
 	c.Assert(err, IsNil)
 
-	// Wait for it to go into backoff state.
-	s.waitUntilService(c, "test2", func(svc *servstate.ServiceInfo) bool {
-		return svc.Current == servstate.StatusBackoff && s.manager.BackoffNum("test2") == 1
-	})
-
-	// Then wait for it to auto-restart (backoff time plus a bit).
-	time.Sleep(75 * time.Millisecond)
-	svc := s.serviceByName(c, "test2")
-	c.Assert(svc.Current, Equals, servstate.StatusActive)
+	// Then wait for it to auto-restart.
+	s.waitForDoneCheck(c, "test2")
+	c.Assert(s.manager.BackoffNum("test2"), Equals, 1)
 	c.Check(s.readAndClearLogBuffer(), Matches, `2.* \[test2\] test2\n`)
 
 	// Send signal to terminate it again.
 	err = s.manager.SendSignal([]string{"test2"}, "SIGTERM")
 	c.Assert(err, IsNil)
 
-	// Wait for it to go into backoff state.
-	s.waitUntilService(c, "test2", func(svc *servstate.ServiceInfo) bool {
-		return svc.Current == servstate.StatusBackoff && s.manager.BackoffNum("test2") == 2
-	})
-
-	// Then wait for it to auto-restart (backoff time plus a bit).
-	time.Sleep(125 * time.Millisecond)
-	svc = s.serviceByName(c, "test2")
-	c.Assert(svc.Current, Equals, servstate.StatusActive)
+	// Then wait for it to auto-restart again.
+	s.waitForDoneCheck(c, "test2")
+	c.Assert(s.manager.BackoffNum("test2"), Equals, 2)
 	c.Check(s.readAndClearLogBuffer(), Matches, `2.* \[test2\] test2\n`)
 
 	// Test that backoff reset time is working (set to backoff-limit)
-	time.Sleep(175 * time.Millisecond)
+	time.Sleep(550 * time.Millisecond)
```

Comment:
Out of interest, why this duration increase?

Reply:
This is a very timing sensitive test, so taking inspiration from your very good question, I've added some comments to the test itself. In summary, the ```backoff-limit``` delay is also used as the backoff reset value. If the test successfully runs for ```backoff-limit``` time, backoff resets. The original value of 150ms meant that a decent cpu spike during the test could easily reset backoff, and the test would fail halfway through. The 550ms value empirically has proven to be more reliable, even under ```stress --cpu 10 --io 10``` usage. Its not technically immune to failures, but statistically has proved way more reliable overall for me. 

The test used to be delay based:

TestActionRestart (delays total = 460ms):
```
:
time.Sleep(10 * time.Millisecond)
:
time.Sleep(75 * time.Millisecond)
:
time.Sleep(125 * time.Millisecond)
:
time.Sleep(175 * time.Millisecond)
:
time.Sleep(75 * time.Millisecond)
:
```

New design: single delay = 550ms

Reply:
And I've changed the test pattern of backoff times:

From: 50ms, 100ms, ... (limit 150ms)
To: 1ms, 10ms, ...         (limit 500ms)

So that a cpu spike on top of the backoff delay is further away from the reset value (backoff limit). I achieved that by using the backoff-factor multiplier (at 10), and a much shorter starting value.

Reply:
Let me know if you are comfortable with this ?

Reply:
Sounds good to me -- thanks for the explanation and comments.